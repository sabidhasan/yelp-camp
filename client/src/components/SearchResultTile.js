import React from 'react'
import RatingBar from './RatingBar'

const SearchResultTile = (props) => {
  return (
    <li className='SearchResultTile'>
      <a
        href={`/campground/${props.id}`}
        onMouseEnter={() => props.onTitleHover(props.id)}
        onMouseLeave={() => props.onTitleHover(null)}
        >
        <h2 className='SearchResultTile__title bold'>{`${props.number}. ${props.name}`}</h2>
      </a>
      <h2 className='SearchResultTile__region'>{props.regionAndProvince}</h2>
      <p className='SearchResultTile__address'>{props.address}</p>
      <p className='SearchResultTile__distance'>{props.distance ? `${props.distance} km away` : null} </p>
      <p className='SearchResultTile__description'>{props.description}</p>
      <img className='SearchResultTile__image' src={props.images ? props.images[0] : null} alt='' />

      <div className='SearchResultTile__rating flex-center'>
        <RatingBar rating={props.rating} small={true} />
        <p className='SearchResultTile__rating-count'>
          {props.comments ? props.comments.length : 0} Reviews
        </p>
      </div>
    </li>
  )
}

export default SearchResultTile
