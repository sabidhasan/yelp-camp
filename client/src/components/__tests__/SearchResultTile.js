import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import SearchResultTile from '../SearchResultTile'

configure({ adapter: new Adapter() });

const generateProps = (key, value) => {
  let tmpProps = {
    id: 4,
    number: 5,
    name: 'Test Name',
    regionAndProvince: 'Region, Province',
    address: '123 Main St, City, PR',
    distance: 10,
    description: 'Some description',
    images: ['source'],
    rating: 5,
    comments: [
      {1: 'one'}, {2: 'one'}, {3: 'one'}, {4: 'one'}
    ]
  }
  if (key) tmpProps[key] = value;
  return tmpProps;
}

describe('<SearchResultTile />', () => {
  let component;
  let props;

  beforeEach(() => {
    const onTitleHover = jest.fn();
    props = { ...generateProps(), onTitleHover }
    component = shallow(<SearchResultTile { ...props } />)
  });

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  });

  it('Distance is not shown, if it is non-existent', () => {
    props['distance'] = undefined;
    component = shallow(<SearchResultTile { ...props } />)
    expect(component.find('.SearchResultTile__distance').text()).toBe('')
  });

  it('If no images supplied, none are shown', () => {
    props['images'] = undefined;
    component = shallow(<SearchResultTile { ...props } />)
    expect(component.find('.SearchResultTile__image').prop('src')).toBe(null)
  });

  it('Proper number of reviews are shown', () => {
    expect(component.find('.SearchResultTile__rating-count').text()).toBe('4 Reviews');
  });

  it('When there is one review, the length is properly shown', () => {
    props['comments'] = [{1: 's'}];
    component = shallow(<SearchResultTile { ...props } />)
    expect(component.find('.SearchResultTile__rating-count').text()).toBe('1 Review');
  });

  it('When there are no reviews, the length is properly shown', () => {
    props['comments'] = undefined;
    component = shallow(<SearchResultTile { ...props } />)
    expect(component.find('.SearchResultTile__rating-count').text()).toBe('No Reviews');
  });

  it('Hovering over title element calls the callback function with correct ID', () => {
    component.find('.SearchResultTile__title-container').simulate('mouseenter')
    expect(props.onTitleHover).toBeCalledWith(4)
  })

  it('Hovering out of title element calls callback function with null', () => {
    component.find('.SearchResultTile__title-container').simulate('mouseleave')
    expect(props.onTitleHover).toBeCalledWith(null)
  })

  it('Snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
})
