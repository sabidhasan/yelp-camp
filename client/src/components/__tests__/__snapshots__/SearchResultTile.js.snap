// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultTile /> Snapshot matches what is expected 1`] = `
<li
  className="SearchResultTile"
>
  <a
    className="SearchResultTile__title-container"
    href="/campground/4"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <h2
      className="SearchResultTile__title bold"
    >
      5. Test Name
    </h2>
  </a>
  <h2
    className="SearchResultTile__region"
  >
    Region, Province
  </h2>
  <p
    className="SearchResultTile__address"
  >
    123 Main St, City, PR
  </p>
  <p
    className="SearchResultTile__distance"
  >
    10 km away
  </p>
  <p
    className="SearchResultTile__description"
  >
    Some description
  </p>
  <img
    alt=""
    className="SearchResultTile__image"
    src="source"
  />
  <div
    className="SearchResultTile__rating flex-center"
  >
    <RatingBar
      rating={5}
      small={true}
    />
    <p
      className="SearchResultTile__rating-count"
    >
      4
       Review
      s
    </p>
  </div>
</li>
`;
